kolibri/plugins/management/assets/src/actions.js

```diff
@@ -69,110 +76,136 @@ function _managePageTitle(title) {
-    // oldRole is admin and role is coach or oldRole is coach and role is admin.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-      // create new role when old role is successfully deleted.
-        const rolePayload = {
-          user: id,
-          collection: FacilityUserModel.attributes.facility,
-          kind: role,
-        };
-        const RoleModel = RoleResource.createModel(rolePayload);
-        RoleModel.save(rolePayload).then((newRole) => {
-        // update the facilityUser when new role is successfully created.
-          FacilityUserModel.save(payload).then(responses => {
-            // force role change because if the role is the only changing attribute
-            // FacilityUserModel.save() will not send request to server.
-            responses.roles = [newRole];
-            store.dispatch('UPDATE_USERS', [responses]);
-          })
-          .catch(error => { coreActions.handleApiError(store, error); });
-        });
-      })
-      .catch(error => { coreActions.handleApiError(store, error); });
-    } else {
-    // role is learner and oldRole is admin or coach.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-        FacilityUserModel.save(payload).then(responses => {
-          // force role change because if the role is the only changing attribute
-          // FacilityUserModel.save() will not send request to server.
-          responses.roles = [];
-          store.dispatch('UPDATE_USERS', [responses]);
-        })
-        .catch(error => { coreActions.handleApiError(store, error); });
+function updateUser(store, stateUser) {
+  // payload needs username, fullname, and facility
+  const userID = stateUser.id;
+  const savedUserModel = FacilityUserResource.getModel(userID);
+  const savedUser = savedUserModel.attributes;
+  const changedValues = {};
+  let roleAssigned = Promise.resolve(savedUserModel.attributes);
+
+  // explicit checks for the only values that can be changed
+  if (stateUser.full_name !== savedUser.full_name) {
+    changedValues.full_name = stateUser.full_name;
+  }
+
+  if (stateUser.username !== savedUser.username) {
+    changedValues.username = stateUser.username;
+  }
+
+  if (stateUser.password !== savedUser.password) {
+    changedValues.password = stateUser.password;
+  }
+
+  if (stateUser.facility !== savedUser.facility) {
+    changedValues.facility = stateUser.facility;
+  }
+
+  if (stateUser.kind !== _userState(savedUser).kind) {
+    // assumes there's no previous roles to delete at first
+    let handlePreviousRoles = Promise.resolve();
+
+    if (savedUser.roles.length) {
+      const roleDeletes = [];
+      savedUser.roles.forEach(role => {
+        roleDeletes.push(RoleResource.getModel(role.id).delete());
       });
+
+
+      // delete the old role models if this was not a learner
+      handlePreviousRoles = Promise.all(roleDeletes).then(responses => {
```

Comment:
This is the pattern I asked about initially, added a comment about in my task list below, and that you mentioned you'd look at later. If there are other changes to be made, I'll take care of them first. This is a reassignment of `165`.

You can see the pattern used again with `144` + `186`

Tried to comment for clarity. I'm already getting the vibe you think this is smelly/too imperative, but I wanted to know for sure before I set out to changing this. Looks like the alternative is either using a `conditionalPromise` or doing conditional additions to an array of promises and having an ultimate `Promise.all`. Might be missing something though.